Director lite

simao opened the pull request:
# Director Rewrite

This is a complete rewrite of director. It follows a few principles to make it simpler that the current director:

-   Simplify the database schema, sometimes at the cost of complex operations at the app level.
-   Only implement features we currently use, based on usage data from production
-   Do not break APIs used by the latest aktualizr
-   Transactionally update a device state when possible

There is still significant work to be done to put this in production, but I wanted to get this rolling and see if we could do it.

This is easier to review if you checkout this branch, or review the branch on github https://github.com/simao/director/tree/director-lite

## OK. But why?

Director has served us well, but it currently one of our main sources of technical debt:

-   It is difficult to change, people are afraid to make any changes to it because almost nobody understands what it's doing.
-   It is doing too much. Director should simply **direct** which update goes to each ecu, everything else should be handled by other services
-   It essentially implements a tuf repository with tuf metadata, but it uses it's own implementation to generate metadata when it could implementations from \`tuf-repo\`
-   It has no owner. Director is currently an orphan, mainly for the above reasons.


## And how?

To be able to keep code as simple as possible, director-lite compromises on a few points:

### [Anti]features

-  No `diff` implementation

    AFAIK, aktualizr does not support this anymore and this introduces a lot of complexity to director. Director should anyway not do any diffing, that should be done by other services.

- ~~No background daemon~~ Background daemon implemented in [OTA-2979](https://saeljira.it.here.com/browse/OTA-2979)

    ~~We can add this later if we need to support the auto update feature or async migrations, but currently there is no reason to have a background service as all role generation is done synchronously, like tuf-repo.~~ 

-  Leverages https://github.com/advancedtelematic/ota-tuf/pull/227 to simplify tuf metadata generation

    A lot of code was re-implemented in director to manage a tuf repository. We already do this and test this in \`tuf-repo\`, we should not re-implement this again.

-  Transactionally processing of device manifest

    This is important and a shortcoming of the current director. When a device pushes a manifest, the database should be updated in a way that stays consistent, even if the user pushes other assignments at the same time.

-  Smaller code base

    This is a smaller code base that the current director. Not only by removing features but also by simplifying the data model, codecs, tests and reusing code from tuf-repo.
    
    This makes director easier to change and fix, and will help encourage other developers to work on director.

-  Only production APIs are implemented

    Director currently has lots of cruft and APIs that are not used by anybody but we still have to maintain and keep because director is so hard to change.
    
    See bellow for details on what is implemented and what isn't.

## And everything works?

No.

Even though I tried to keep the API used by aktualizr intact, it's very likely that I introduced some regressions there, but they should be fixable and we should be able to make the latest aktualizr versions work.

However, I did break the API that devices do not use, so the UI would need some changes.

I analysed production logs over a few weeks to find out which director APIs are being used in production and tried to keep as much as I could here.

Production Hits |        |                                                       | Implemented          | Notes                          
---------------- | ------ | ----------------------------------------------------- | -------------------- | -------------------------------
68827 | PUT    | /api/v1/device/:uuid/manifest                         | Y                    |                                
64679 | GET    | /api/v1/device/:uuid/root.json                        | Y                    |                                
64651 | GET    | /api/v1/device/:uuid/targets.json                     | Y                    |                                
1514 | GET    | /api/v1/admin/devices/:uuid                           | Y                    |                                
99 | GET    | /api/v1/admin/devices/:uuid/ecus/public_key           | Y (Endpoint Changed) | ../ecus/:ecu_serial/public_key 
85 | POST   | /api/v1/admin/images/installed_count                  | Y                    |                                
75 | GET    | /api/v1/admin/devices/:uuid/queue                     | N                    | use assignments                
42 | GET    | /api/v1/admin/devices/:uuid/ecus/:ecu-id/auto_update  | Y                   |                    
36 | POST   | /api/v1/device/:uuid/ecus                             | Y                    |                                
23 | GET    | /api/v1/device/:uuid/1.root.json                      | Y                    |                                
11 | POST   | /api/v1/multi_target_updates                          | Y                    |                                
9 | PUT    | /api/v1/admin/devices/:uuid/multi_target_update/:uuid | N                    | use assignments                
5 | GET    | /api/v1/multi_target_updates/:uuid                    | Y                    |                                
3 | POST   | /api/v1/assignments                                   | Y (Endpoint Changed) | GET ../assignments/devices     
3 | GET    | /api/v1/admin/devices/hardware_identifiers            | Y                    |                                
3 | GET    | /api/v1/admin/devices                                 | N                    | Use device registry?           
3 | DELETE | /api/v1/assignments/:uuid                             | N          | Use PATCH     
2 | GET    | /api/v1/device/:uuid/2.root.json                      | Y                    |                                
1 | PATCH  | /api/v1/assignments                                   | Y          |     

Other than this, there are a few things missing before we can deploy this to production:


### Migrating data from old director

This can be done, but it's complicated and I wanted to wait until I had some feedback before I implemented it.


### ~~Backwards compatibility to old aktualizr~~
Implemented by [OTA-2980](https://saeljira.it.here.com/browse/OTA-2980)

~~The only problem with this, after we migrate the data, is supporting the old json format that older aktualizr versions are using.~~

~~One way to do this would be to keep the old director running just for the old devices. If we want to avoid this, we could also introduce a simple layer that converts a legacy manifest to a new manifest for the new director to process, instead of adding complexity to the codecs implementation.~~

### ~~Cancel Assignment~~ 
Implemented by [OTA-2983](https://saeljira.it.here.com/browse/OTA-2983)
~~This is not implemented but should be possible to implement.~~

### Auto Update
~~IMO this should not be done by director, but other than that there is no reason we could not implement it here.~~

Implemented in [OTA-2979](https://saeljira.it.here.com/browse/OTA-2979)


### No support for moving keys offline

IMO there is no use case for this, director always needs to have targets keys online, and I see no security benefit of having the root keys offline for director.


### No Root Role Rotation

This needs to be implemented but we can reuse code from `tuf-repo`.

# Feedback

Any feedback is welcome, please let me know if you think we should continue with this, or if I am wrong, or this is just stupid.

simao committed:
Remove diff service

simao committed:
Re write director

simao committed:
device gets logged when fetching root with version

Commit:
Bump sbt and scalac versions

Commit:
Assignment cancelation

Signed-off-by: Jochen Schneider <j.schneider@here.com>

simao committed:
Merge pull request #1 from jochenschneider/feat/OTA-2983/assignment-cancelation

Feat/ota 2983/assignment cancelation

simao committed:
implement auto update

Signed-off-by: Simão Mata <simao.mata@here.com>

Commit:
Process old device manifest versions

Signed-off-by: Jochen Schneider <j.schneider@here.com>

Commit:
Merge pull request #188 from advancedtelematic/feat/OTA-2980/device-manifest-compatibility

Process old device manifest versions

zabbal commented:
> No support for moving keys offline

Do we have this documented somewhere? Is it used in production?

Commit:
Idempotent migration director1 -> director2

Signed-off-by: Jochen Schneider <j.schneider@here.com>

Commit:
Fixes and optimizations

Signed-off-by: Jochen Schneider <j.schneider@here.com>

Commit:
Scala migration for director1.filecache to director2.signed_roles

Signed-off-by: Jochen Schneider <j.schneider@here.com>

Commit:
Merge pull request #192 from advancedtelematic/feat/OTA-3901/migrate-signed_roles

Scala migration for director1.filecache to director2.signed_roles

Commit:
Merge pull request #190 from advancedtelematic/feat/OTA-3901/migration

Idempotent migration director1 -> director2

simao committed:
remove unused settings

Signed-off-by: Simão Mata <simao.mata@here.com>

simao committed:
Added migration for all needed tables from director to lite

Moved data migration out of flyway, at least for now. It looks like
this migration would be better if executed manually where we could go
step by step and confirm the counts along the way.

Signed-off-by: Simão Mata <simao.mata@here.com>

simao committed:
add missing indeces, fix slow queries

Signed-off-by: Simão Mata <simao.mata@here.com>

simao committed:
Merge pull request #196 from simao/director-lite

Added migration for all needed tables from director to lite
